crates/bevy_tasks/src/executor.rs

```diff
@@ -0,0 +1,598 @@
+        }
+    }
+
+    /// Waits for the next runnable task to run, given a function that searches for a task.
+    async fn runnable_with(&self, mut search: impl FnMut() -> Option<Runnable>) -> Runnable {
+        future::poll_fn(|cx| {
+            loop {
+                match search() {
+                    None => {
+                        // Move to sleeping and unnotified state.
+                        if !self.sleep(cx.waker()) {
+                            // If already sleeping and unnotified, return.
+                            return Poll::Pending;
+                        }
+                    }
+                    Some(r) => {
+                        // Wake up.
+                        self.wake();
+
+                        // Notify another ticker now to pick up where this ticker left off, just in
+                        // case running the task takes a long time.
+                        self.state.notify();
+
+                        return Poll::Ready(r);
+                    }
+                }
+            }
+        })
+        .await
+    }
+}
+
+impl Drop for Ticker<'_> {
+    fn drop(&mut self) {
+        // If this ticker is in sleeping state, it must be removed from the sleepers list.
+        let id = self.sleeping.swap(0, Ordering::SeqCst);
+        if id != 0 {
+            let mut sleepers = self.state.sleepers.lock();
+            let notified = sleepers.remove(id);
+
+            self.state
+                .notified
+                .swap(sleepers.is_notified(), Ordering::SeqCst);
+
+            // If this ticker was notified, then notify another ticker.
+            if notified {
+                drop(sleepers);
+                self.state.notify();
+            }
+        }
+    }
+}
+
+/// A worker in a work-stealing executor.
+///
+/// This is just a ticker that also has an associated local queue for improved cache locality.
+#[derive(Debug)]
+struct Runner<'a> {
+    priority: usize,
+
+    /// The executor state.
+    state: &'a State,
+
+    /// Inner ticker.
+    ticker: Ticker<'a>,
+
+    /// The local queue.
+    local: Arc<ConcurrentQueue<Runnable>>,
+
+    /// Bumped every time a runnable task is found.
+    ticks: AtomicUsize,
+}
+
+impl Runner<'_> {
+    /// Creates a runner and registers it in the executor state.
+    fn new(priority: usize, thread_id: usize, state: &State) -> Runner<'_> {
+        let local = state.local_queues[priority][thread_id].clone();
+        let runner = Runner {
+            priority,
+            state,
+            ticker: Ticker::new(state),
+            local, // Arc::new(ConcurrentQueue::bounded(512)),
+            ticks: AtomicUsize::new(0),
+        };
+        runner
+    }
+
+    fn priority_iter(&self) -> impl Iterator<Item = usize> {
+        // Prioritize the immediate responsibility of the runner, then search in reverse order
+        std::iter::once(self.priority).chain((self.priority + 1..self.state.queues.len()).rev())
```

Comment:
this doesn't look right to me 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=95825c6c49b51eff179672ee43c84824

in the above playground it doesn't ever iterate over zero. also don't we want the ordering to be more like 1,0,2,3,4?

Reply:
oh wait nvm. I guess this is the intended behavior. i.e. we only steal tasks from higher priority pools.

Reply:
We *shouldn't* attempt to steal from lower priority queues. As stated in the description, we must prevent lower priority tasks from starving out higher priority ones, which means threads cannot work steal from even lower priority threads, as that risks blocking higher priority tasks.